File: app/src/main/java/com/shop440/Receiver/SmsReciever.java
package com.shop440.Receiver;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.telephony.SmsMessage;
import android.util.Log;

import com.shop440.Models.SmsListener;

public class SmsReciever extends BroadcastReceiver {
    private static SmsListener smsListener;

    @Override
    public void onReceive(Context context, Intent intent) {
    Bundle data = intent.getExtras();
        Log.d("Reciever", "test");
        Object[] pdus = (Object[]) data.get("pdus");
        for(int i=0; i<pdus.length; i++){
            SmsMessage smsMessage = SmsMessage.createFromPdu((byte[]) pdus[i]);
            String sender = smsMessage.getDisplayOriginatingAddress();
            if(sender.equals("shop440")){
                String messageBody = smsMessage.getMessageBody();
                smsListener.messageReceived(messageBody);
            }
        }
    }

    public static void bindListener(SmsListener listener) {
        smsListener = listener;
    }
}

File: app/src/main/java/com/shop440/receiver/SmsReciever.kt
package com.shop440.receiver

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.telephony.SmsMessage
import android.util.Log
import com.shop440.utils.SmsListener

class SmsReciever : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        val data = intent.extras
        Log.d("Reciever", "test")
        val pdus = data!!.get("pdus") as Array<Any>
        for (i in pdus.indices) {
            val smsMessage = SmsMessage.createFromPdu(pdus[i] as ByteArray)
            val sender = smsMessage.displayOriginatingAddress
            if (sender == "shop440") {
                val messageBody = smsMessage.messageBody
                smsListener!!.messageReceived(messageBody)
            }
        }
    }

    companion object {
        private var smsListener: SmsListener? = null

        fun bindListener(listener: SmsListener) {
            smsListener = listener
        }
    }
}
